site/content/blog/2019-11-04-unit-testing-mmctl-commands.md

```diff
@@ -0,0 +1,144 @@
+
+# Asserting the Command's Behavior
+
+After mocking the server interactions, all that's left to do is run the command and check the outputs. There are two things that we can use to check that the command behaved as we expected: the command's return value and whatever was printed during its execution.
+
+To perform these assertions we can use the suite itself for [assertions](https://godoc.org/github.com/stretchr/testify/assert) or `s.Require()` for [requires](https://godoc.org/github.com/stretchr/testify/require). Both implement the same helpers to check that our values meet our expectations. The difference between the two is that a failed check with `require` will mark the test as failed and will stop it right after, and a failed check with `assert` will mark the test as failed too, but it will continue running it.
+
+The rule of thumb here is to use `assert` when you can still get valuable information from the following assertions of your test, and use `require` if, after an error, the following checks won't make sense. For example, if we get an error opening a file, it doesn't make sense to check the file contents.
+
+This is what the command execution and its assertions look like:
+
+```go
+err := searchUserCmdF(s.client, &cobra.Command{}, []string{emailArg})
+s.Require().Nil(err)
+s.Require().Equal(&mockUser, printer.GetLines()[0])
+s.Require().Len(printer.GetErrorLines(), 0)
+```
+
+In the case of searching for an existing user, the first thing we expect is the error to come back as `nil`, so that's our first thing to assert.
+
+To check the output of the commands, we use the `printer` struct. Every significant message that a command prints goes through this struct, which accumulates the output during the execution. At testing time, we can use these accumulated lines to check that the output matches our expectations.
+
+We have two kinds of output: the lines for `stdout` and the error lines for `stderr`. In our case, as the user should have been found and printed through `stdout`, we can use the `printer.GetLines` method to get a slice with the printed messages, and the `Equal` helper to perform the assertion. Then, as we are expecting no errors, we can check that the `printer` struct printed none during the command execution asserting that `printer.GetErrorLines` is empty.
+
+# Printer Cleanup
+
+One thing we need to remember is that although the `printer` struct gets cleaned for us between test functions courtesy of `MmctlUnitTestSuite`, we will need to clean it manually between test cases. Therefore, if we have more than one use case in the same test function, we need to run `printer.Clean()` at the beginning of all the `s.Run` blocks but the first one:
+
+```go
+func (s *MmctlUnitTestSuite) TestMyCommandCmd() {
```

Comment:
I'd be tempted to just recommend we have `printer.Clean()` in every test function, that way it doesn't lead to weirdness if someone inserts a new test at the start of a function rather than the end in the future.

Reply:
Sounds good, I've modified the paragraph and the example to reflect the change :)